tyrian/src/main/scala/tyrian/cmds/FileReader.scala

```diff
@@ -16,103 +18,88 @@ import scala.scalajs.js.typedarray
-        try Result.File(n, p, IArray.from(new typedarray.Int8Array(d.asInstanceOf[typedarray.ArrayBuffer])))
-        catch case _ => Result.Error("Could not cast loaded file data to byte array")
-
-    readFromInputField(inputFieldId, ReadType.AsArrayBuffer)(cast andThen resultToMessage)
+    readFile(getFileFromInput(inputFieldId), ReadType.AsArrayBuffer)(readBytesCast andThen resultToMessage)
 
   /** Reads an input file as bytes */
   def readBytes[F[_]: Async, Msg](file: dom.File)(resultToMessage: Result[IArray[Byte]] => Msg): Cmd[F, Msg] =
-    val cast: Result[js.Any] => Result[IArray[Byte]] =
-      case Result.Error(msg) => Result.Error(msg)
-      case Result.File(n, p, d) =>
-        try Result.File(n, p, IArray.from(new typedarray.Int8Array(d.asInstanceOf[typedarray.ArrayBuffer])))
-        catch case _ => Result.Error("Could not cast loaded file data to byte array")
-
-    readFile(file, ReadType.AsArrayBuffer)(cast andThen resultToMessage)
+    readFile(Some(file).pure[F], ReadType.AsArrayBuffer)(readBytesCast andThen resultToMessage)
 
-  private def readFromInputField[F[_]: Async, Msg](fileInputFieldId: String, readType: ReadType)(
-      resultToMessage: Result[js.Any] => Msg
-  ): Cmd[F, Msg] =
-    val files = document.getElementById(fileInputFieldId).asInstanceOf[html.Input].files
-    if files.length == 0 then Cmd.None
-    else readFile(files.item(0), readType)(resultToMessage)
+  private def getFileFromInput[F[_]: Async](inputFieldId: String): F[Option[dom.File]] =
+    Async[F].delay(document.getElementById(inputFieldId).asInstanceOf[html.Input].files.headOption)
 
-  private def readFile[F[_]: Async, Msg](file: dom.File, readAsType: ReadType)(
+  private def readFile[F[_]: Async, Msg](maybeGetFile: F[Option[dom.File]], readAsType: ReadType)(
       resultToMessage: Result[js.Any] => Msg
   ): Cmd[F, Msg] =
-    val task = Async[F].delay {
-      val p          = Promise[Result[js.Any]]()
-      val fileReader = new dom.FileReader()
-      fileReader.addEventListener(
-        "load",
-        (e: Event) =>
-          p.success(
-            Result.File(
-              name = file.name,
-              path = readAsType match
-                case ReadType.AsDataUrl => e.target.asInstanceOf[js.Dynamic].result.asInstanceOf[String]
-                case _                  => ""
-              ,
-              data = fileReader.result
-            )
-          ),
-        false
-      )
-      fileReader.onerror = _ => p.success(Result.Error(s"Error reading from file"))
-
-      readAsType match
-        case ReadType.AsText => fileReader.readAsText(file)
-        case ReadType.AsArrayBuffer =>
-          fileReader.readAsArrayBuffer(file)
-        case ReadType.AsDataUrl => fileReader.readAsDataURL(file)
-
-      p.future
+    val task = maybeGetFile.flatMap {
+      case None => Future.successful(Result.NoFile("No files on specified input")).pure[F]
+      case Some(file) =>
+        Async[F].delay {
```

Comment:
thought: the sequence of `callback -> promise -> future -> f` seems like it should be `Async[F].async`. On top of the conversions having extra cost, it doesn't handle cancellation (which might result in memory leaking the event listener/promise in case the command gets cancelled).

It would probably be beneficial to rewrite this as `Async[F].async` and remove the event listener in case of interruption. note, I can't remember if command effects actually are interruptible.

Reply:
> I can't remember if command effects actually are interruptible

In practice, not really. You can do it but you have to work hard to make it happen, it isn't just easily available. It is one of those topics that comes up a lot on the backend but... my hypothesis is that it's less important on the frontend, and that if you have one of those cases where it really does matter then you won't mind doing the work.

So it's a tradeoff. In this arch style, precise effect management is more difficult, but the overall code and life cycle is (IMO) simpler and cleaner (Discussions on scaling not withstanding :stuck_out_tongue:).

Reply:
I agree that there is a good argument for converting to `Async[F].async`, however I don't think that's @sloshy's problem. This code is an improvement, so no need to hold up the PR any longer.

But it's a good suggest, and I've [captured it as an issue](https://github.com/PurpleKingdomGames/tyrian/issues/280).